runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java

```diff
@@ -401,14 +401,25 @@ public MetricResults metrics() {
      * {@link DirectOptions#isShutdownUnboundedProducersWithMaxWatermark()} set to false,
      * this method will never return.
      *
-     * <p>See also {@link PipelineExecutor#awaitCompletion()}.
+     * <p>See also {@link PipelineExecutor#awaitCompletion(Duration)}.
      */
     @Override
     public State waitUntilFinish() {
-      if (!state.isTerminal()) {
+      return waitUntilFinish(Duration.ZERO);
+    }
+
+    @Override
+    public State cancel() {
+      executor.stop();
+      return waitUntilFinish();
```

Comment:
This logic shouldn't really be present at all - this dispatches a call to the executor to shut it down, and then returns immediately. If the proper behavior includes waiting to finish, the call to `waitToFinish` should be explicit.

`stop` will stop the executor from starting any more work, but need not immediately shut down any in-progress work (but should work towards turning down any active work)